webapp/tests/karma/ts/services/rules-engine.service.spec.ts

```diff
@@ -0,0 +1,619 @@
+    fetchTargetsResult = sinon.stub().resolves([]);
+    const clock = sinon.useFakeTimers(1000);
+    service = TestBed.inject(RulesEngineService);
+
+    await service.fetchTargets();
+    clock.tick(500 * 1000);
+    await service.isEnabled(); // to resolve promises
+
+    expect(rulesEngineCoreStubs.fetchTasksFor.callCount).to.eq(1);
+    expect(rulesEngineCoreStubs.fetchTargets.callCount).to.eq(1);
+    expect(telemetryService.record.callCount).to.equal(5);
+    expect(telemetryService.record.args[0][0]).to.equal('rules-engine:initialize');
+    expect(telemetryService.record.args[1][0]).to.equal('rules-engine:targets:dirty-contacts');
+    expect(telemetryService.record.args[2][0]).to.equal('rules-engine:ensureTargetFreshness:cancel');
+    expect(telemetryService.record.args[3][0]).to.equal('rules-engine:targets');
+    expect(telemetryService.record.args[4][0]).to.equal('rules-engine:tasks:dirty-contacts');
+  });
+
+  it('should cancel all ensure freshness threads', async () => {
+    fetchTargetsResult = sinon.stub().resolves([]);
+    fetchTasksResult = sinon.stub().resolves([]);
+    const clock = sinon.useFakeTimers(1000);
+    service = TestBed.inject(RulesEngineService);
+
+    service.fetchTaskDocsForAllContacts();
+    await service.fetchTargets();
+    clock.tick(500 * 1000);
+    await service.isEnabled(); // to resolve promises
+    
+    expect(rulesEngineCoreStubs.fetchTasksFor.callCount).to.eq(1);
+    expect(rulesEngineCoreStubs.fetchTargets.callCount).to.eq(1);
+    expect(telemetryService.record.callCount).to.equal(7);
+    expect(telemetryService.record.args[0][0]).to.equal('rules-engine:initialize');
+    expect(telemetryService.record.args[1][0]).to.equal('rules-engine:tasks:dirty-contacts');
+    expect(telemetryService.record.args[2][0]).to.equal('rules-engine:ensureTaskFreshness:cancel');
+    expect(telemetryService.record.args[3][0]).to.equal('rules-engine:targets:dirty-contacts');
+    expect(telemetryService.record.args[4][0]).to.equal('rules-engine:ensureTargetFreshness:cancel');
+    expect(telemetryService.record.args[5][0]).to.equal('rules-engine:tasks:all-contacts');
+    expect(telemetryService.record.args[6][0]).to.equal('rules-engine:targets');
+  });
+
+  it('should record correct telemetry data with emitted events', async () => {
+    const realSetTimeout = setTimeout;
+    const nextTick = () => new Promise(resolve => realSetTimeout(() => resolve()));
+    const clock = sinon.useFakeTimers(1000);
+    let fetchTargetResultPromise;
+    const fetchTasksResultPromise = [];
+    fetchTargetsResult = sinon.stub().callsFake(() => new Promise(resolve => fetchTargetResultPromise = resolve));
+    fetchTasksResult = sinon.stub().callsFake(() => new Promise(resolve => fetchTasksResultPromise.push(resolve)));
+    service = TestBed.inject(RulesEngineService);
+
+    service.fetchTargets();
+    service.fetchTaskDocsForAllContacts();
+    service.fetchTaskDocsFor(['a']);
+    await service.isEnabled();
```

Comment:
I think the original test called `isEnabled` before `fetchTargets` and all these other calls.

Reply:
If I call `isEnabled` before other calls, like the original test did, assertions fail.

Reply:
Fixed